File: tools/experiments/shared-semantics/lua/shared-semantics.lua
-- Experimental same-source-Lua control. This is deliberately isolated from
-- production. The exact same file is executed by LÖVE/LuaJIT and Wasmoon/Lua 5.4.
local shared = {}

local ASSET_DIRS = {
    "assets/smallBattlers",
    "assets/sprites",
    "assets/system",
}

local function copyTokens(tokens)
    local out = {}
    for k, v in pairs(tokens or {}) do out[k] = v end
    return out
end

function shared.parseSpriteKey(spriteKey)
    local tokens = {}
    local fileKey = tostring(spriteKey):gsub("%[([^=]+)=([^%]]+)%]", function(k, v)
        tokens[k] = tonumber(v) or v
        return ""
    end)
    fileKey = fileKey:gsub("^%s*(.-)%s*$", "%1")
    return { fileKey = fileKey, tokens = tokens }
end

local function mergedTokens(keyTokens, filenameTokens)
    local merged = copyTokens(filenameTokens)
    for k, v in pairs(keyTokens or {}) do merged[k] = v end
    return merged
end

function shared.resolveTiming(keyTokens, filenameTokens)
    keyTokens = keyTokens or {}
    filenameTokens = filenameTokens or {}
    local merged = mergedTokens(keyTokens, filenameTokens)
    if merged.fps ~= nil then
        local numeric = tonumber(merged.fps)
        return {
            fps = numeric,
            source = keyTokens.fps ~= nil and "key"
                or (filenameTokens.fps ~= nil and "filename" or "resolved"),
            token = "fps",
            value = merged.fps,
        }
    elseif merged.speed ~= nil then
        local numeric = tonumber(merged.speed)
        return {
            fps = numeric and 4 * numeric or nil,
            source = keyTokens.speed ~= nil and "key"
                or (filenameTokens.speed ~= nil and "filename" or "resolved"),
            token = "speed",
            value = merged.speed,
        }
    end
    return { fps = 4, source = "default", token = nil, value = nil }
end

local function containsPath(files, path)
    for _, candidate in ipairs(files or {}) do
        if candidate == path then return true end
    end
    return false
end

local function basename(path)
    return tostring(path):match("([^/\\]+)$") or tostring(path)
end

local function titleCaseFirst(value)
    if #value == 0 then return value end
    return value:sub(1, 1):upper() .. value:sub(2):lower()
end

local function stripPng(filename)
    return filename:gsub("%.png$", "")
end

local function indexedFilename(files, fileKey)
    local wanted = fileKey:lower()
    for _, dir in ipairs(ASSET_DIRS) do
        local prefix = dir .. "/"
        for _, path in ipairs(files or {}) do
            if path:sub(1, #prefix) == prefix then
                local localName = path:sub(#prefix + 1)
                if not localName:find("/", 1, true) and localName:match("%.png$") then
                    local parsed = shared.parseSpriteKey(stripPng(localName))
                    if parsed.fileKey:lower() == wanted then
                        return { path = path, tokens = parsed.tokens }
                    end
                end
            end
        end
    end
    return nil
end

function shared.resolveSpriteKey(spriteKey, files)
    local parsed = shared.parseSpriteKey(spriteKey)
    local fileKey = parsed.fileKey
    local indexed = indexedFilename(files, fileKey)
    local filenameTokens = indexed and copyTokens(indexed.tokens) or {}
    local candidates = {
        "assets/smallBattlers/" .. titleCaseFirst(fileKey) .. ".png",
        "assets/smallBattlers/" .. fileKey .. ".png",
        "assets/smallBattlers/" .. fileKey:lower() .. ".png",
        "assets/sprites/" .. fileKey .. ".png",
        "assets/system/" .. fileKey .. ".png",
        "assets/system/" .. titleCaseFirst(fileKey) .. ".png",
    }
    local path
    for _, candidate in ipairs(candidates) do
        if containsPath(files, candidate) then
            path = candidate
            break
        end
    end
    if not path and indexed then path = indexed.path end
    return {
        resolved = path ~= nil,
        key = spriteKey,
        path = path,
        tokenSourcePath = indexed and indexed.path or nil,
        keyTokens = copyTokens(parsed.tokens),
        filenameTokens = filenameTokens,
        tokens = mergedTokens(parsed.tokens, filenameTokens),
        timing = shared.resolveTiming(parsed.tokens, filenameTokens),
    }
end

function shared.describeSpritePath(path)
    local parsed = shared.parseSpriteKey(stripPng(basename(path)))
    local filenameTokens = parsed.tokens
    return {
        resolved = true,
        key = "",
        path = path,
        tokenSourcePath = path,
        keyTokens = {},
        filenameTokens = copyTokens(filenameTokens),
        tokens = copyTokens(filenameTokens),
        timing = shared.resolveTiming({}, filenameTokens),
    }
end

local MODULUS = 65521
local HASH_MULTIPLIER = 25173
local HASH_ADDEND = 13849
local MAX_SEED = 2147483646
local FRACTAL_PERSISTENCE = 0.55
local FRACTAL_OCTAVES = {
    { 0.8, -0.6, 0.6, 0.8, 3.17, -5.29 },
    { 0.6, 0.8, -0.8, 0.6, 17.17, -9.31 },
    { -0.8, 0.6, -0.6, -0.8, -13.73, 21.47 },
    { -0.6, -0.8, 0.8, -0.6, 29.11, 14.53 },
}

local function positiveModulo(value, modulus)
    local result = value % modulus
    if result < 0 then result = result + modulus end
    return result
end

local function lerp(a, b, t)
    return a + (b - a) * t
end

local function smoothstep(t)
    return t * t * (3 - 2 * t)
end

function shared.hash01(x, y, seed)
    local ix = positiveModulo(math.floor(x), MODULUS)
    local iy = positiveModulo(math.floor(y), MODULUS)
    local iseed = positiveModulo(math.floor(seed or 0), MODULUS)
    local value = (ix * 3749 + iy * 9151 + iseed * 1013) % MODULUS
    value = (value * HASH_MULTIPLIER + HASH_ADDEND) % MODULUS
    value = (value * HASH_MULTIPLIER + HASH_ADDEND) % MODULUS
    return value / (MODULUS - 1)
end

function shared.valueNoise(x, y, seed)
    local x0, y0 = math.floor(x), math.floor(y)
    local fx, fy = x - x0, y - y0
    local sx, sy = smoothstep(fx), smoothstep(fy)
    local top = lerp(shared.hash01(x0, y0, seed), shared.hash01(x0 + 1, y0, seed), sx)
    local bottom = lerp(shared.hash01(x0, y0 + 1, seed), shared.hash01(x0 + 1, y0 + 1, seed), sx)
    return lerp(top, bottom, sy)
end

function shared.fractalNoise(x, y, seed)
    local total, amplitude, normalizer, frequency = 0, 1, 0, 1
    for octaveIndex, octave in ipairs(FRACTAL_OCTAVES) do
        local rotatedX = (x * octave[1] + y * octave[2] + octave[5]) * frequency
        local rotatedY = (x * octave[3] + y * octave[4] + octave[6]) * frequency
        total = total + shared.valueNoise(rotatedX, rotatedY, seed + (octaveIndex - 1) * 7919) * amplitude
        normalizer = normalizer + amplitude
        amplitude = amplitude * FRACTAL_PERSISTENCE
        frequency = frequency * 2
    end
    return total / normalizer
end

local function validateRgb(problems, value, where)
    if type(value) ~= "table" or #value ~= 3 then
        problems[#problems + 1] = where .. " must be an RGB triple"
        return
    end
    for channel = 1, 3 do
        if type(value[channel]) ~= "number" or value[channel] < 0 or value[channel] > 1 then
            problems[#problems + 1] = where .. " channel " .. channel .. " must be a number in 0..1"
        end
    end
end

local function isDenseList(value)
    if type(value) ~= "table" then return false end
    local count = 0
    for key in pairs(value) do
        if type(key) ~= "number" or key < 1 or key ~= math.floor(key) then return false end
        count = count + 1
    end
    return count == #value
end

function shared.validate(layers, where)
    where = where or "vertexShadingLayers"
    local problems = {}
    if layers == nil then return problems end
    if not isDenseList(layers) then
        problems[#problems + 1] = where .. " must be a dense list"
        return problems
    end
    for index, layer in ipairs(layers) do
        local desc = where .. "[" .. index .. "]"
        if type(layer) ~= "table" then
            problems[#problems + 1] = desc .. " must be an object"
        elseif layer.type ~= "colorNoise" then
            problems[#problems + 1] = desc .. ".type '" .. tostring(layer.type)
                .. "' is unsupported (expected colorNoise)"
        else
            validateRgb(problems, layer.colorA, desc .. ".colorA")
            validateRgb(problems, layer.colorB, desc .. ".colorB")
            if type(layer.strength) ~= "number" or layer.strength < 0 or layer.strength > 1 then
                problems[#problems + 1] = desc .. ".strength must be a number in 0..1"
            end
            if type(layer.scale) ~= "number" or layer.scale <= 0 then
                problems[#problems + 1] = desc .. ".scale must be a number > 0"
            end
            if type(layer.seed) ~= "number" or layer.seed ~= math.floor(layer.seed)
                    or math.abs(layer.seed) > MAX_SEED then
                problems[#problems + 1] = desc .. ".seed must be an integer between -"
                    .. MAX_SEED .. " and " .. MAX_SEED
            end
        end
    end
    return problems
end

function shared.compile(layers, where)
    local problems = shared.validate(layers, where)
    if #problems > 0 then error(table.concat(problems, "\n"), 0) end
    local compiled = {}
    for _, layer in ipairs(layers or {}) do
        compiled[#compiled + 1] = {
            type = layer.type,
            colorA = { layer.colorA[1], layer.colorA[2], layer.colorA[3] },
            colorB = { layer.colorB[1], layer.colorB[2], layer.colorB[3] },
            strength = layer.strength,
            scale = layer.scale,
            seed = layer.seed,
        }
    end
    return compiled
end

function shared.sampleCompiled(compiled, x, y)
    local r, g, b = 1, 1, 1
    for _, layer in ipairs(compiled or {}) do
        local noise = shared.fractalNoise(x / layer.scale, y / layer.scale, layer.seed)
        local nr = lerp(layer.colorA[1], layer.colorB[1], noise)
        local ng = lerp(layer.colorA[2], layer.colorB[2], noise)
        local nb = lerp(layer.colorA[3], layer.colorB[3], noise)
        r = r * lerp(1, nr, layer.strength)
        g = g * lerp(1, ng, layer.strength)
        b = b * lerp(1, nb, layer.strength)
    end
    return r, g, b
end

function shared.sample(layers, x, y)
    return shared.sampleCompiled(shared.compile(layers), x, y)
end

function shared.grid(layers, width, height)
    local compiled = shared.compile(layers)
    local out = {}
    for y = 0, height do
        local row = {}
        out[y + 1] = row
        for x = 0, width do
            local r, g, b = shared.sampleCompiled(compiled, x, y)
            row[x + 1] = { r, g, b }
        end
    end
    return out
end

return shared

File: tools/experiments/shared-semantics/src/shared-semantics.ts
/*
 * Experimental shared semantic source. This file is intentionally NOT wired to
 * production runtime or Studio. It is compiled twice: ordinary JavaScript via
 * tsc, and LuaJIT-target Lua via TypeScriptToLua.
 */
namespace ThestraSharedSemantics {
    export type TokenValue = number | string;
    export interface TokenMap { [key: string]: TokenValue; }

    export interface ParsedSpriteKey {
        fileKey: string;
        tokens: TokenMap;
    }

    export interface SpriteTiming {
        fps: number | null;
        source: 'key' | 'filename' | 'default' | 'resolved';
        token: 'fps' | 'speed' | null;
        value: TokenValue | null;
    }

    export interface SpriteResolution {
        resolved: boolean;
        key: string;
        path: string | null;
        tokenSourcePath: string | null;
        keyTokens: TokenMap;
        filenameTokens: TokenMap;
        tokens: TokenMap;
        timing: SpriteTiming;
    }

    const ASSET_DIRS = [
        'assets/smallBattlers',
        'assets/sprites',
        'assets/system'
    ];

    function isAsciiWhitespace(code: number): boolean {
        return code === 32 || code === 9 || code === 10 || code === 11 || code === 12 || code === 13;
    }

    function trimAscii(value: string): string {
        let first = 0;
        let last = value.length;
        while (first < last && isAsciiWhitespace(value.charCodeAt(first))) first++;
        while (last > first && isAsciiWhitespace(value.charCodeAt(last - 1))) last--;
        return value.substring(first, last);
    }

    function tokenValue(raw: string): TokenValue {
        const trimmed = trimAscii(raw);
        if (trimmed.length === 0) return raw;
        const numeric = Number(trimmed);
        return Number.isFinite(numeric) ? numeric : raw;
    }

    function copyTokens(tokens: TokenMap): TokenMap {
        const result: TokenMap = {};
        for (const key in tokens) result[key] = tokens[key];
        return result;
    }

    export function parseSpriteKey(spriteKey: string): ParsedSpriteKey {
        const tokens: TokenMap = {};
        let fileKey = '';
        let cursor = 0;

        while (cursor < spriteKey.length) {
            const open = spriteKey.indexOf('[', cursor);
            if (open < 0) {
                fileKey += spriteKey.substring(cursor);
                break;
            }
            fileKey += spriteKey.substring(cursor, open);
            const equals = spriteKey.indexOf('=', open + 1);
            const close = equals >= 0 ? spriteKey.indexOf(']', equals + 1) : -1;
            if (equals > open + 1 && close > equals + 1) {
                const key = spriteKey.substring(open + 1, equals);
                const value = spriteKey.substring(equals + 1, close);
                tokens[key] = tokenValue(value);
                cursor = close + 1;
            } else {
                fileKey += '[';
                cursor = open + 1;
            }
        }

        return { fileKey: trimAscii(fileKey), tokens };
    }

    function mergedTokens(keyTokens: TokenMap, filenameTokens: TokenMap): TokenMap {
        const merged = copyTokens(filenameTokens);
        for (const key in keyTokens) merged[key] = keyTokens[key];
        return merged;
    }

    function numericToken(value: TokenValue): number | null {
        if (typeof value === 'number') return Number.isFinite(value) ? value : null;
        const trimmed = trimAscii(value);
        if (trimmed.length === 0) return null;
        const numeric = Number(trimmed);
        return Number.isFinite(numeric) ? numeric : null;
    }

    export function resolveTiming(keyTokens: TokenMap, filenameTokens: TokenMap): SpriteTiming {
        const merged = mergedTokens(keyTokens, filenameTokens);
        if (merged.fps !== undefined) {
            const value = merged.fps;
            const numeric = numericToken(value);
            return {
                fps: numeric,
                source: keyTokens.fps !== undefined ? 'key'
                    : (filenameTokens.fps !== undefined ? 'filename' : 'resolved'),
                token: 'fps',
                value
            };
        }
        if (merged.speed !== undefined) {
            const value = merged.speed;
            const numeric = numericToken(value);
            return {
                fps: numeric === null ? null : 4 * numeric,
                source: keyTokens.speed !== undefined ? 'key'
                    : (filenameTokens.speed !== undefined ? 'filename' : 'resolved'),
                token: 'speed',
                value
            };
        }
        return { fps: 4, source: 'default', token: null, value: null };
    }

    function containsPath(files: string[], path: string): boolean {
        for (let index = 0; index < files.length; index++) {
            if (files[index] === path) return true;
        }
        return false;
    }

    function basename(path: string): string {
        let slash = -1;
        for (let index = 0; index < path.length; index++) {
            const code = path.charCodeAt(index);
            if (code === 47 || code === 92) slash = index;
        }
        return path.substring(slash + 1);
    }

    function lower(value: string): string { return value.toLowerCase(); }

    function titleCaseFirst(value: string): string {
        if (value.length === 0) return value;
        return value.substring(0, 1).toUpperCase() + value.substring(1).toLowerCase();
    }

    function stripPng(filename: string): string {
        return lower(filename.substring(filename.length - 4)) === '.png'
            ? filename.substring(0, filename.length - 4)
            : filename;
    }

    function indexedFilename(files: string[], fileKey: string): { path: string; tokens: TokenMap } | null {
        const wanted = lower(fileKey);
        for (let dirIndex = 0; dirIndex < ASSET_DIRS.length; dirIndex++) {
            const dir = ASSET_DIRS[dirIndex];
            const prefix = dir + '/';
            for (let fileIndex = 0; fileIndex < files.length; fileIndex++) {
                const path = files[fileIndex];
                if (path.substring(0, prefix.length) !== prefix) continue;
                const localName = path.substring(prefix.length);
                if (localName.indexOf('/') >= 0 || lower(localName.substring(localName.length - 4)) !== '.png') continue;
                const parsed = parseSpriteKey(stripPng(localName));
                if (lower(parsed.fileKey) === wanted) return { path, tokens: parsed.tokens };
            }
        }
        return null;
    }

    export function resolveSpriteKey(spriteKey: string, files: string[]): SpriteResolution {
        const parsed = parseSpriteKey(spriteKey);
        const keyTokens = parsed.tokens;
        const directPaths = [
            'assets/smallBattlers/' + titleCaseFirst(parsed.fileKey) + '.png',
            'assets/smallBattlers/' + parsed.fileKey + '.png',
            'assets/smallBattlers/' + lower(parsed.fileKey) + '.png',
            'assets/sprites/' + parsed.fileKey + '.png',
            'assets/system/' + parsed.fileKey + '.png',
            'assets/system/' + titleCaseFirst(parsed.fileKey) + '.png'
        ];
        const indexed = indexedFilename(files, parsed.fileKey);
        const filenameTokens = indexed ? copyTokens(indexed.tokens) : {};
        const merged = mergedTokens(keyTokens, filenameTokens);

        let path: string | null = null;
        for (let index = 0; index < directPaths.length; index++) {
            if (containsPath(files, directPaths[index])) {
                path = directPaths[index];
                break;
            }
        }
        if (path === null && indexed) path = indexed.path;

        return {
            resolved: path !== null,
            key: spriteKey,
            path,
            tokenSourcePath: indexed ? indexed.path : null,
            keyTokens: copyTokens(keyTokens),
            filenameTokens,
            tokens: merged,
            timing: resolveTiming(keyTokens, filenameTokens)
        };
    }

    export function describeSpritePath(path: string): SpriteResolution {
        const parsed = parseSpriteKey(stripPng(basename(path)));
        const filenameTokens = parsed.tokens;
        return {
            resolved: true,
            key: '',
            path,
            tokenSourcePath: path,
            keyTokens: {},
            filenameTokens: copyTokens(filenameTokens),
            tokens: copyTokens(filenameTokens),
            timing: resolveTiming({}, filenameTokens)
        };
    }

    const MODULUS = 65521;
    const HASH_MULTIPLIER = 25173;
    const HASH_ADDEND = 13849;
    const MAX_SEED = 2147483646;
    const FRACTAL_PERSISTENCE = 0.55;
    const FRACTAL_OCTAVES = [
        [0.8, -0.6, 0.6, 0.8, 3.17, -5.29],
        [0.6, 0.8, -0.8, 0.6, 17.17, -9.31],
        [-0.8, 0.6, -0.6, -0.8, -13.73, 21.47],
        [-0.6, -0.8, 0.8, -0.6, 29.11, 14.53]
    ];

    export interface ShadingLayer {
        type: string;
        colorA: number[];
        colorB: number[];
        strength: number;
        scale: number;
        seed: number;
    }

    function positiveModulo(value: number, modulus: number): number {
        const result = value % modulus;
        return result < 0 ? result + modulus : result;
    }

    function lerp(a: number, b: number, t: number): number {
        return a + (b - a) * t;
    }

    function smoothstep(t: number): number {
        return t * t * (3 - 2 * t);
    }

    export function hash01(x: number, y: number, seed: number): number {
        const ix = positiveModulo(Math.floor(x), MODULUS);
        const iy = positiveModulo(Math.floor(y), MODULUS);
        const iseed = positiveModulo(Math.floor(seed || 0), MODULUS);
        let value = (ix * 3749 + iy * 9151 + iseed * 1013) % MODULUS;
        value = (value * HASH_MULTIPLIER + HASH_ADDEND) % MODULUS;
        value = (value * HASH_MULTIPLIER + HASH_ADDEND) % MODULUS;
        return value / (MODULUS - 1);
    }

    export function valueNoise(x: number, y: number, seed: number): number {
        const x0 = Math.floor(x);
        const y0 = Math.floor(y);
        const fx = x - x0;
        const fy = y - y0;
        const sx = smoothstep(fx);
        const sy = smoothstep(fy);
        const top = lerp(hash01(x0, y0, seed), hash01(x0 + 1, y0, seed), sx);
        const bottom = lerp(hash01(x0, y0 + 1, seed), hash01(x0 + 1, y0 + 1, seed), sx);
        return lerp(top, bottom, sy);
    }

    export function fractalNoise(x: number, y: number, seed: number): number {
        let total = 0;
        let amplitude = 1;
        let normalizer = 0;
        let frequency = 1;
        for (let octaveIndex = 0; octaveIndex < FRACTAL_OCTAVES.length; octaveIndex++) {
            const octave = FRACTAL_OCTAVES[octaveIndex];
            const rotatedX = (x * octave[0] + y * octave[1] + octave[4]) * frequency;
            const rotatedY = (x * octave[2] + y * octave[3] + octave[5]) * frequency;
            total += valueNoise(rotatedX, rotatedY, seed + octaveIndex * 7919) * amplitude;
            normalizer += amplitude;
            amplitude *= FRACTAL_PERSISTENCE;
            frequency *= 2;
        }
        return total / normalizer;
    }

    function validateRgb(problems: string[], value: number[], where: string): void {
        if (value == null || typeof value !== 'object' || value.length !== 3) {
            problems.push(where + ' must be an RGB triple');
            return;
        }
        for (let channel = 0; channel < 3; channel++) {
            const sample = value[channel];
            if (typeof sample !== 'number' || !Number.isFinite(sample) || sample < 0 || sample > 1) {
                problems.push(where + ' channel ' + (channel + 1) + ' must be a number in 0..1');
            }
        }
    }

    export function validate(layers: ShadingLayer[] | null | undefined, where = 'vertexShadingLayers'): string[] {
        const problems: string[] = [];
        if (layers == null) return problems;
        if (typeof layers !== 'object') {
            problems.push(where + ' must be a dense list');
            return problems;
        }
        for (let index = 0; index < layers.length; index++) {
            const layer = layers[index];
            const desc = where + '[' + (index + 1) + ']';
            if (layer == null || typeof layer !== 'object') {
                problems.push(desc + ' must be an object');
            } else if (layer.type !== 'colorNoise') {
                problems.push(desc + ".type '" + String(layer.type) + "' is unsupported (expected colorNoise)");
            } else {
                validateRgb(problems, layer.colorA, desc + '.colorA');
                validateRgb(problems, layer.colorB, desc + '.colorB');
                if (typeof layer.strength !== 'number' || !Number.isFinite(layer.strength)
                        || layer.strength < 0 || layer.strength > 1) {
                    problems.push(desc + '.strength must be a number in 0..1');
                }
                if (typeof layer.scale !== 'number' || !Number.isFinite(layer.scale) || layer.scale <= 0) {
                    problems.push(desc + '.scale must be a number > 0');
                }
                if (typeof layer.seed !== 'number' || !Number.isFinite(layer.seed)
                        || Math.floor(layer.seed) !== layer.seed || Math.abs(layer.seed) > MAX_SEED) {
                    problems.push(desc + '.seed must be an integer between -' + MAX_SEED + ' and ' + MAX_SEED);
                }
            }
        }
        return problems;
    }

    export function compile(layers: ShadingLayer[] | null | undefined, where = 'vertexShadingLayers'): ShadingLayer[] {
        const problems = validate(layers, where);
        if (problems.length > 0) throw new Error(problems.join('\n'));
        const compiled: ShadingLayer[] = [];
        if (layers == null) return compiled;
        for (let index = 0; index < layers.length; index++) {
            const layer = layers[index];
            compiled.push({
                type: layer.type,
                colorA: [layer.colorA[0], layer.colorA[1], layer.colorA[2]],
                colorB: [layer.colorB[0], layer.colorB[1], layer.colorB[2]],
                strength: layer.strength,
                scale: layer.scale,
                seed: layer.seed
            });
        }
        return compiled;
    }

    export function sampleCompiled(compiled: ShadingLayer[] | null | undefined, x: number, y: number): number[] {
        let r = 1;
        let g = 1;
        let b = 1;
        if (compiled != null) {
            for (let index = 0; index < compiled.length; index++) {
                const layer = compiled[index];
                const noise = fractalNoise(x / layer.scale, y / layer.scale, layer.seed);
                const nr = lerp(layer.colorA[0], layer.colorB[0], noise);
                const ng = lerp(layer.colorA[1], layer.colorB[1], noise);
                const nb = lerp(layer.colorA[2], layer.colorB[2], noise);
                r *= lerp(1, nr, layer.strength);
                g *= lerp(1, ng, layer.strength);
                b *= lerp(1, nb, layer.strength);
            }
        }
        return [r, g, b];
    }

    export function sample(layers: ShadingLayer[] | null | undefined, x: number, y: number): number[] {
        return sampleCompiled(compile(layers), x, y);
    }

    export function grid(layers: ShadingLayer[] | null | undefined, width: number, height: number): number[][][] {
        const compiled = compile(layers);
        const result: number[][][] = [];
        for (let y = 0; y <= height; y++) {
            const row: number[][] = [];
            for (let x = 0; x <= width; x++) row.push(sampleCompiled(compiled, x, y));
            result.push(row);
        }
        return result;
    }
}
